ugvc/mrd/featuremap_consensus_utils.py

```diff
@@ -0,0 +1,111 @@
+    record_dict["VAF"] = record_dict["FILTERED_COUNT"] / record_dict["DP"]
+    rec.filter.add(curr_filter)
+
+    for info_field in info_fields:
+        rec.info[info_field] = record_dict[info_field]
+
+    record_dict["GT"] = (0, 1)
+    record_dict["AD"] = [record_dict["FILTERED_COUNT"], record_dict["DP"] - record_dict["FILTERED_COUNT"]]
+    for format_field in format_fields:
+        rec.samples["SAMPLE"][format_field] = record_dict[format_field]
+
+    out_fh.write(rec)
+    return rec
+
+
+def pileup_featuremap(featuremap: str, output_dir: str, genomic_interval: str, min_qual: int):
+    """
+    Pileup featuremap vcf to a reagular vcf, save the mean quality and read count
+    """
+    cons_dict = defaultdict(dict)
+    # process genomic interval input
+    if genomic_interval == "":
+        chrom = None
+        start = None
+        end = None
+    if genomic_interval != "":
+        genomic_interval_list = genomic_interval.split(":")
+        chrom = genomic_interval_list[0]
+        start = int(genomic_interval_list[1].split("-")[0])
+        end = int(genomic_interval_list[1].split("-")[1])
+
+    # generate a new header
+    orig_header = pysam.VariantFile(featuremap).header
+    header = pysam.VariantHeader()
+    header.add_meta("fileformat", "VCFv4.2")
+    header.filters.add("FAIL", None, None, "Mean quality below threshold")
+    header.info.add(
+        "FILTERED_COUNT",
+        1,
+        "Integer",
+        "Number of reads containing this location that agree with alternative according to fitler",
+    )
+    header.info.add("X_READ_COUNT", 1, "Integer", "Number of reads containing this location")
+    header.info.add("X_QUAL", ".", "Float", "Quality of reads containing this location")
+    header.info.add("X_RN", 1, "String", "Read Name of reads containing this location, entries spearated by |")
+    header.info.add("X_AF", 1, "Float", "Allele frequency")
+    header.formats.add("GT", 1, "String", "Genotype")
+    header.formats.add("DP", 1, "Integer", "Read depth")
+    header.formats.add("AD", "R", "Integer", "Allelic depths")
+    header.formats.add("VAF", 1, "Float", "Allele frequency")
+    # copy contigs from original header
+    for contig in orig_header.contigs:
+        header.contigs.add(contig)
+    header.add_sample("SAMPLE")
+
+    # open an output file
+    output_vcf = pjoin(
+        output_dir, basename(featuremap).replace(".vcf.gz", f".{genomic_interval.replace(':', '-')}.pileup.vcf.gz")
+    )
+    out_fh = pysam.VariantFile(output_vcf, "w", header=header)
+
+    cons_dict = defaultdict(dict)
+    with pysam.VariantFile(featuremap) as f:
+        for rec in f.fetch(chrom, start, end):
+            rec_id = "-".join([str(x) for x in (rec.chrom, rec.pos, rec.ref, rec.alts[0])])
+            if "FILTERED_COUNT" not in cons_dict[rec_id]:
+                if len(cons_dict.keys()) > 1:
+                    # write to file
+                    prev_key = list(cons_dict.keys())[0]
```

Comment:
I'm concerned about this line, because I don't think the order of keys is guaranteed. I think you need an OrderedDict for that. 

Also, is it possible you are counting the last entry twice? You require "len(cons_dict.keys()) > 1" before you assign cons_dict[rec_id], so I think you will do another iteration of the lines that add to cons_dict[rec_id] before entering the if-condition that pops the prev_key.

Reply:
I think I solved it now